src/main/kotlin/three/consulting/epoc/controller/CustomerController.kt

```diff
@@ -15,4 +15,7 @@ class CustomerController(private val customerService: CustomerService) {
 
     @PostMapping()
     fun createCustomer(@Valid @RequestBody customer: CustomerDTO) = customerService.createCustomer(customer)
+
+    @PutMapping(value = ["/{customerId}"])
+    fun updateCustomerForId(@PathVariable customerId: Long, @Valid @RequestBody customer: CustomerDTO) = customerService.updateCustomerForId(customerId, customer)
```

Comment:
The separate id parameter is not needed. the request body already contains it. Only validation in controller should be if the name is null or blank.

Reply:
This was something a should have discussed with you earlier. As the current solution is quite incoherent. I have thought that PUT requests should always be addressed to a single entity (eq. here `/customer/{id}`) compared to POST which is newer addressed to a single entity ( eq. here `/customer` ). What is your take on this?

Currently, the problem is that the id must be given in the URL, but the payload id can be different and the change is made according to the payload id, and the id given in URL is neglected. So it is possible to send a PUT request to URL /customer/1 to change the values of the customer with id 2. I assume this is unwanted behavior.

Which is the preferable way to solve this out:
 1. Remove the id from the URL
 2.  Check that ids match in the controller level
 3. Check that ids match in the service level
Or is there a different approach?

Reply:
I'd drop the id as pathvariable as it's a duplicate and as you said it just causes more prone for error. From controller just pass the request body to service layer and there check if id is null or not. Just like the post mapping and creating a customer in service layer, just with the id check logic reversed.

Reply:
I'll use path variable where the only thing needed is a single id not containing whitespace. So getting one customer and in the next issue deleting one can be done with the id and no body is required. But I never mix a path variable and a request body in the same controller method, it just makes code harder to read and leaves more room for error